chapters/part3/hyperparameter-optimization-with-optuna.md

```diff
@@ -0,0 +1,419 @@
+      ngram_filter_sizes: ngram_filter_sizes,
+      num_filters: num_filters,
+      output_dim: output_dim,
+    },
+    dropout: dropout,
+  },
+  data_loader: {
+    batch_size: batch_size,
+  },
+  trainer: {
+    cuda_device: cuda_device,
+    num_epochs: num_epochs,
+    optimizer: {
+      lr: lr,
+      type: 'sgd',
+    },
+    validation_metric: '+accuracy',
+  },
+}
+```
+
+</details>
+
+## II: Defining Search Space for Hyperparameters in Pythonic way
+
+Now that you have created the config, the next step is defining the hyperparameter search spaces.
+In Optuna, a search space is defined by creating an `objective function`.
+Each hyperparameter search space is declared with `suggest_int` or `suggest_float`.
+For categorical hyperparameters, you can use `suggest_categorical`.
+Please see <a href="https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial">Optuna documentation</a> for more information.
+
+These suggest functions require two kinds of arguments at least.
+The first one is the name of hyperparameter, and the second one is the range of the values.
+Note that the names of the hyperparameters should be the same as those defined in the configuration earlier.
+A typical objective function looks like following:
+
+```python
+import optuna
+
+
+def objective(trial: optuna.Trial) -> float:
+    trial.suggest_int("embedding_dim", 32, 256)
+    trial.suggest_int("max_filter_size", 2, 6)
+    trial.suggest_int("num_filters", 32, 256)
+    trial.suggest_int("output_dim", 32, 256)
+    trial.suggest_float("dropout", 0.0, 0.8)
+    trial.suggest_float("lr", 5e-3, 5e-1, log=True)
+
+    executor = optuna.integration.allennlp.AllenNLPExecutor(
+        trial=trial,  # trial object
+        config_file="./config/imdb_optuna.jsonnet",  # path to jsonnet
+        serialization_dir=f"./result/optuna/{trial.number}",
+        metrics="best_validation_accuracy"
+    )
+    return executor.run()
+```
+
+After defining search spaces using `trial.suggest_int` and `trial.suggest_float`, the `trial` object should be passed to `AllenNLPExecutor`.
+The `trial` object holds all suggested values after defining search spaces using `trial.suggest_int` and `trial.suggest_float`,  and it should be pass to `AllenNLPExecutor`.
+`AllenNLPExecutor` takes four required arguments; `trial` (Optuna's object), `config_file` (path to a model configuration),
+`serialization_dir` (directory for saving model snapshot, log, etc.), and `metrics` you want to optimize.
+In the above example, we create an instance of `AllenNLPExecutor` as `executor`.
+Once the `executor` instance is created, training is started with `executor.run()`.
+In each trial step in optimization, the objective function is called and does the following steps:
+
+1. Train a model (`executor.run()`)
+2. Return a target metric on validation data (`executor.run()` returns the specified metric)
+
+Now, we finished defining objective function. :tada:
```

Comment:
In my environment, `:tada:` was not rendered properly. I guess we need [`gatsby-remark-emojis`](https://www.gatsbyjs.org/packages/gatsby-remark-emojis/).

![image](https://user-images.githubusercontent.com/3255979/89896693-db03b300-dc18-11ea-8537-ec497baa1db7.png)

Reply:
Thank you for catching, I didn't notice it.
I removed emoji in a711c66.